clash-prelude/src/Clash/Sized/Fixed.hs

```diff
@@ -904,6 +905,182 @@ fLitR a = Fixed (fromInteger sat)
     truncated = truncate shifted :: Integer
     shifted   = a * (2 ^ (natToInteger @frac))
 
+-- | These behave similar to 'Prelude.Float', 'Prelude.Double' and
+-- 'Prelude.Rational'. 'succ'\/'pred' add\/subtract 1. See the
+-- <https://www.haskell.org/onlinereport/haskell2010/haskellch6.html#dx13-131001 Haskell Report>
+-- for full details.
+--
+-- The rules set out there for instances of both 'Enum' and
+-- 'Bounded' are also observed. In particular, 'succ' and 'pred' result in a
+-- runtime error if the result cannot be represented. See 'satSucc' and
+-- 'satPred' for other options.
+instance NumFixedC rep int frac => Enum (Fixed rep int frac) where
+  succ f =
+    let err = error $
+             "Enum.succ{" ++ show (typeOf f) ++ "}: tried to take 'succ' of "
+          ++ show f ++ ", causing overflow. Use 'satSucc' and specify a "
+          ++ "SaturationMode if you need other behavior."
+    in case natToInteger @int of
+         0 -> err
+         _ -> if f > satPred SatBound maxBound then
+                err
+              else
+                satSucc SatWrap f
+
+
+  pred f =
+    let err = error $
+             "Enum.pred{" ++ show (typeOf f) ++ "}: tried to take 'pred' of "
+          ++ show f ++ ", causing negative overflow. Use 'satPred' and "
+          ++ "specify a SaturationMode if you need other behavior."
+    in case natToInteger @int of
+         0 -> err
+         _ -> if f < satSucc SatBound minBound then
+                err
+              else
+                satPred SatWrap f
+
+  toEnum i =
+    let sh   = natToNum @frac
+        res  = toInteger i `shiftL` sh
+        rMax = toInteger (maxBound :: rep (int + frac))
+        rMin = toInteger (minBound :: rep (int + frac))
+    in if res > rMax || res < rMin then
+         error $  "Enum.toEnum{"
+               ++ show (typeRep $ Proxy @(Fixed rep int frac)) ++ "}: tag ("
+               ++ show i ++ ") is outside of bounds "
+               ++ show ( minBound :: Fixed rep int frac
+                       , maxBound :: Fixed rep int frac)
+       else
+         Fixed (fromInteger res)
```

Comment:
If you want to prevent all the indention whitespace, consider:

```haskell
  toEnum i =
    if res > rMax || res < rMin then
      error $  "Enum.toEnum{"
             ++ show (typeRep $ Proxy @(Fixed rep int frac)) ++ "}: tag ("
             ++ show i ++ ") is outside of bounds "
             ++ show ( minBound :: Fixed rep int frac
                     , maxBound :: Fixed rep int frac)
    else
       Fixed (fromInteger res)
   where
    sh   = natToNum @frac
    res  = toInteger i `shiftL` sh
    rMax = toInteger (maxBound :: rep (int + frac))
    rMin = toInteger (minBound :: rep (int + frac))
```

Especially in larger blocks, it makes it easier to see how many control structures deep you are IMO.

Reply:
We should really add something like `Clash.Util.Interpolate` to `clash-prelude`. Writing error messages like this:

https://github.com/clash-lang/clash-compiler/blob/250aea43ce6111cd55cdb83e1ad30761024c9bb0/clash-lib/src/Clash/Core/TermInfo.hs#L185-L193

is so much easier to read (and write)!

Reply:
I've transformed the `let` to `where` so the indentation is less.